align/pnr/placer_pythonic.py

```diff
@@ -30,37 +30,35 @@ def trim_placement_data(placement_data, top_level):
     modules = {module['concrete_name']: module for module in placement_data['modules']}
 
     top_concrete_names = [module['concrete_name'] for module in placement_data['modules'] if module['abstract_name'] == top_level]
-    all_modules_leaves = list()
+    all_modules_leaves = set()
     for concrete_name in top_concrete_names:
         _, found_modules, found_leaves = compute_topoorder(modules, concrete_name, key='concrete_template_name')
-        all_modules_leaves.extend(found_modules)
-        all_modules_leaves.extend(found_leaves)
-
-    all_modules_leaves = set(all_modules_leaves)
+        all_modules_leaves.update(found_modules)
+        all_modules_leaves.update(found_leaves)
 
     new_placement_data = {'leaves': list(), 'modules': list()}
-    new_placement_data['leaves'] = [leaf for leaf in placement_data['leaves'] if leaf['concrete_name'] in all_modules_leaves]
-    new_placement_data['modules'] = [leaf for leaf in placement_data['modules'] if leaf['concrete_name'] in all_modules_leaves]
-
     for key in ['leaves', 'modules']:
+        new_placement_data[key] = [x for x in placement_data[key] if x['concrete_name'] in all_modules_leaves]
         for elem in new_placement_data[key]:
             if 'pin_bbox' in elem:
                 del elem['pin_bbox']
+            if 'global_signals' in elem:
+                elem['global_signals'] = list(elem['global_signals'])
 
     return new_placement_data
 
 
 def propagate_down_global_signals(modules: dict, module_name: str, global_signals: list):
     GS = 'global_signals'
-    modules[module_name][GS] = modules[module_name].get(GS, list()) + global_signals
+    modules[module_name][GS] = set.union(modules[module_name].get(GS, set()), global_signals)
```

Comment:
I'm not sure this helps with potential quadratic runtime.
The set.union will build a whole new set which takes O(n) time. You want it to take time proportional to the size of global_signals.

Maybe:
```
if GS in modules[module_name]:
   modules[module_name][GS].update(global_signals)
else:
  modules[module_name][GS] = set(global_signals)
```